core/arch/arm/mm/core_mmu.c

```diff
@@ -887,10 +936,12 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 #endif
 			map->attr = core_mmu_type_to_attr(map->type);
 			va = ROUNDUP(va, map->region_size);
-#if !defined(CFG_WITH_LPAE)
-			/* Mapping does not yet support sharing L2 tables */
-			va = ROUNDUP(va, CORE_MMU_PGDIR_SIZE);
-#endif
+			/*
+			 * Make sure that va is aligned with pa for
+			 * efficient pgdir mapping. Basically pa &
+			 * pgdir_mask should be == va & pgdir_mask
+			 */
+			va += (map->pa - va) & CORE_MMU_PGDIR_MASK;
```

Comment:
Removing the old code (va roundup in case of non LPAE) is ok.
But regarding the added code, I feel this may over consume mmu entries. It seems useful only if the area spreads at least over a superblock. And even so, since it shifts the other areas, it is not obvious there is gain here. 
But maybe i failed to understand your change...

Reply:
Ah, yes, my first version checked for `map->size >= PGDIR_SIZE * 2`. But I didn't noticed any changes in tables usage. I can return it back, if you feel that it is needed.

Reply:
You can simply force the case by hacking this way:
```patch
--- a/core/arch/arm/mm/core_mmu.c
+++ b/core/arch/arm/mm/core_mmu.c
@@ -898,7 +898,7 @@ static void init_mem_map(struct tee_mmap_region *memory_map, size_t num_elems)
 
        assert(!((va | end) & SMALL_PAGE_MASK));
 
-       if (core_mmu_place_tee_ram_at_top(va)) {
+       if (1 || core_mmu_place_tee_ram_at_top(va)) {
                /* Map non-flat mapped addresses below flat mapped addresses */
                for (map = memory_map; !core_mmap_is_end_of_table(map); map++) {
                        if (map->va)
```